### PR TITLE
Introduce and generate libsodium bindings using rust-bindgen

### PR DESCRIPTION
**Pull request by cynecx**
The transition to rust-bindgen actually went really smooth, besides some "issues" like the hardcoded `crypto_pwhash_scryptsalsa208sha256_STRPREFIX` constant, everything else seems trivial.

Most of the changes were:

- Use `as_ptr()` and `as_mut_ptr()` instead of relying on rust-ized fitting types on the bindings part
- bindgen generates constants with `u32` as type, hence all the type casts to `usize`
- While at it, rustfmt the tests in libsodium-sys
- Moved libsodium-sys tests into their own modules
- Introduced new env variable: `SODIUM_INC_DIR`

### CODE REVIEW

File: libsodium-sys/Cargo.toml [hunk 1]
```diff
@@ -1,25 +1,25 @@
...
 categories = ["cryptography", "api-bindings"]
-version = "0.0.16"
+version = "0.0.17"
 
 [badges]
 travis-ci = { repository = "dnaq/sodiumoxide" }
 
 [build-dependencies]
 pkg-config = "0.3"
+bindgen = { git = "https://github.com/cynecx/rust-bindgen", branch = "canonicalize_pointer_ty" }
```
Comment (Dylan-DPC) [inline libsodium-sys/Cargo.toml]: Waiting on https://github.com/rust-lang-nursery/rust-bindgen/pull/1311 so that we don't have to depend on a fork

File: src/crypto/scalarmult/curve25519.rs [hunk 2]
```diff
@@ -26,12 +26,13 @@ new_type! {
 /// `Ok(q)`.
 /// If the the `GroupElement` is all zero, `scalarmult()` returns `Err(())` since
 /// the resulting `GroupElement` would be all zero, no matter the `Scalar`.
-pub fn scalarmult(&Scalar(ref n): &Scalar,
-                  &GroupElement(ref p): &GroupElement)
-                  -> Result<GroupElement, ()> {
+pub fn scalarmult(
```
Comment (dnaq) [inline src/crypto/scalarmult/curve25519.rs]: There are a lot of whitespace changes here that complicate reviewing the diff, especially since we have a bunch of test vectors here, would be better if this file didn't have these whitespace changes.

Comment (kinggoesgaming) [reply src/crypto/scalarmult/curve25519.rs]: I agree with this... maybe introduce a separate PR before this PR that runs the `rustfmt`?

File: libsodium-sys/build.rs [hunk 3]
```diff
@@ -1,24 +1,40 @@
...
-        println!("cargo:rustc-link-lib={0}=sodium", mode);
 
+        println!("cargo:rustc-link-lib={0}=sodium", mode);
     } else {
-
         pkg_config::find_library("libsodium").unwrap();
-
     }
 
+    let include_dir = pkg_config::get_variable("libsodium", "includedir").unwrap();
```
[the remark this answers is not in the document]
Comment (cynecx) [reply libsodium-sys/build.rs]: Oh crap, I've totally missed that part while trying to fix the CI.
So if we are using `SODIUM_LIB_DIR`, we can imply the include dir with `${SODIUM_LIB_DIR}/../include`? Or do we have to introduce another variable?

File: libsodium-sys/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,59 @@
+#![allow(non_upper_case_globals)]
```
[the remark this answers is not in the document]
Comment (cynecx) [reply libsodium-sys/src/lib.rs]: I agree, I will apply the changes asap.

File: libsodium-sys/build.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dnaq) [reply libsodium-sys/build.rs]: Maybe we should introduce another variable so that we don't make any assumptions on the build system?

File: libsodium-sys/Cargo.toml (hunk 1 above)
Comment (cynecx) [reply libsodium-sys/Cargo.toml]: It seems that the follow up RP has been merged as well. Waiting for that version bump :)

Comment (kinggoesgaming) [reply libsodium-sys/Cargo.toml]: `bindgen@0.37.0` has been released.

**Comment by cynecx** [comment]
Alright, we are using upstream bindgen now and the feature-branch should be rebased.

File: libsodium-sys/Cargo.toml [hunk 5]
```diff
@@ -1,25 +1,25 @@
 [package]
 authors = ["dnaq"]
 build = "build.rs"
-description = "FFI binding to libsodium"
+description = "FFI binding to libsodium generated with bindgen"
```
Comment (kinggoesgaming) [inline libsodium-sys/Cargo.toml]: I don't think we need to specify that its bindgen generated in the description

File: libsodium-sys/Cargo.toml [hunk 6]
```diff
@@ -1,25 +1,25 @@
...
 documentation = "https://sodiumoxide.github.io/sodiumoxide"
 keywords = ["libsodium", "NaCl", "crypto"]
 license = "MIT OR Apache-2.0"
 links = "sodium"
 name = "libsodium-sys"
 repository = "https://github.com/sodiumoxide/sodiumoxide.git"
 categories = ["cryptography", "api-bindings"]
-version = "0.0.16"
+version = "0.0.17"
```
Comment (kinggoesgaming) [inline libsodium-sys/Cargo.toml]: This bump should ideally be its own PR, just as to keep the merge conflicts out of the way

Comment (cynecx) [reply libsodium-sys/Cargo.toml]: I agree.

File: libsodium-sys/Cargo.toml (hunk 5 above)
Comment (cynecx) [reply libsodium-sys/Cargo.toml]: Yup.

**Comment by cynecx** [comment]
Rebased again. I think squashing the commits will be the last step until the review is complete.

**Comment by Dylan-DPC** [comment]
You don't have to rebase manually. There is an "update branch" button at the end of the PR which will do it for you

**Comment by cynecx** [comment]
@Dylan-DPC Tbh, I've never used that feature :) How does it deal with merge conflicts or is it even possible to skip commits? (`git rebase --skip`)

**Comment by cynecx** [comment]
Is there anything left todo?

File: libsodium-sys/build.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Dylan-DPC) [reply libsodium-sys/build.rs]: @cynecx is this done?

**Comment by Dylan-DPC** [comment]
> Tbh, I've never used that feature :) How does it deal with merge conflicts or is it even possible to skip commits? 

@cynecx if there are conflicts it shows you which files are conflicting and you can resolve it in the Github UI editor or do it normally through code. Not sure it can skip commits. 

If you are done with the PR you can remove the `[WIP]` from the title. Once we get approval from @kinggoesgaming or @dnaq or anyone else we can merge this

**Comment by cynecx** [comment]
@dnaq I am actually thinking that we should address this in another PR since this one is already "packed" with code changes. It would make review process much easier imo.

**Comment by dnaq** [comment]
@cynecx Agreed
bors r+
